spectral_cube/cube_utils.py

```diff
@@ -784,17 +795,65 @@ def combine_headers(header1, header2, **kwargs):
     wcs_opt, shape_opt = find_optimal_celestial_wcs([(s1, w1), (s2, w2)], auto_rotate=False,
                                                     **kwargs)
 
+    # find spectral coverage
+    specw1 = WCS(header1).spectral
+    specw2 = WCS(header2).spectral
+    specaxis1 = [x[0] for x in WCS(header1).world_axis_object_components].index('spectral')
+    specaxis2 = [x[0] for x in WCS(header2).world_axis_object_components].index('spectral')
+    range1 = specw1.pixel_to_world([0,header1[f'NAXIS{specaxis1+1}']-1])
+    range2 = specw2.pixel_to_world([0,header2[f'NAXIS{specaxis1+1}']-1])
+
+    # check for overlap
+    # this will raise an exception if the headers are an different units, which we want
+    if max(range1) < min(range2) or max(range2) < min(range1):
+        warnings.warn(f"There is no spectral overlap between {range1} and {range2}")
+
+    # check cdelt
+    dx1 = specw1.proj_plane_pixel_scales()[0]
+    dx2 = specw2.proj_plane_pixel_scales()[0]
+    if np.abs((dx1 - dx2)/dx1) > spectral_dx_threshold:
+        raise ValueError(f"Different spectral pixel scale {dx1} vs {dx2}")
+
+    ranges = np.hstack([range1, range2])
+    new_naxis = int(np.ceil((ranges.max() - ranges.min()) / np.abs(dx1)))
+    if specw1.wcs.cdelt == 1.0:
+        raise NotImplementedError("Spectral WCS doesn't have a CDELT parameter; we don't know how to deal with this generally.")
+    if np.sign(specw1.wcs.cdelt) == 1:
+        new_crval3 = ranges.min().to(u.Hz).value
+    elif np.sign(specw1.wcs.cdelt) == -1:
+        new_crval3 = ranges.max().to(u.Hz).value
+    else:
+        raise "WTF?"
+
     # Make a new header using the optimal wcs and information from cubes
     header = header1.copy()
     header['NAXIS'] = 3
     header['NAXIS1'] = shape_opt[1]
     header['NAXIS2'] = shape_opt[0]
-    header['NAXIS3'] = header1['NAXIS3']
+    header['NAXIS3'] = new_naxis
     header.update(wcs_opt.to_header())
+    header['CRVAL3'] = new_crval3
     header['WCSAXES'] = 3
```

Comment:
```suggestion
    header['WCSAXES'] = 3
    header['BITPIX'] = header1['BITPIX']
```

Set `BITPIX` explicitly here? I was running into cases where BITPIX was dropped in spatial slices:

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[10], line 1
----> 1 out = mosaic_cubes(cubes,
      2                    spectral_block_size=100,
      3                    combine_header_kwargs={},
      4                    target_header=None,
      5                 #    commonbeam=None,
      6                    commonbeam=Beam(12 * u.arcsec),
      7                    weightcubes=None,
      8                    save_to_tmp_dir=True,
      9                    use_memmap=True,
     10                    output_file=str(output_filename),
     11                 #    method='channel',
     12                    method='cube',
     13                    verbose=True,
     14                    fail_if_cube_dropped=False,
     15                    fail_if_channel_empty=False,)

File ~/Library/CloudStorage/Dropbox/code_development/radio_astro_tools/spectral-cube/spectral_cube/cube_utils.py:930, in mosaic_cubes(cubes, spectral_block_size, combine_header_kwargs, target_header, commonbeam, weightcubes, save_to_tmp_dir, use_memmap, output_file, method, verbose, fail_if_cube_dropped, fail_if_channel_empty, **kwargs)
    928 # Prepare an array and mask for the final cube
    929 shape_opt = (target_header['NAXIS3'], target_header['NAXIS2'], target_header['NAXIS1'])
--> 930 nbits = int(abs(target_header['BITPIX']))
    931 if nbits not in (16, 32, 64):
    932     raise ValueError(f"Invalid BITPIX specified: {nbits} (must be 16, 32, or 64)")

File ~/anaconda3/envs/linearmosaicking-dev/lib/python3.10/site-packages/astropy/io/fits/header.py:170, in Header.__getitem__(self, key)
    167 else:
    168     keyword = key
--> 170 card = self._cards[self._cardindex(key)]
    172 if card.field_specifier is not None and keyword == card.rawkeyword:
    173     # This is RVKC; if only the top-level keyword was specified return
    174     # the raw value, not the parsed out float value
    175     return card.rawvalue

File ~/anaconda3/envs/linearmosaicking-dev/lib/python3.10/site-packages/astropy/io/fits/header.py:1771, in Header._cardindex(self, key)
   1768         indices = self._rvkc_indices.get(keyword, None)
   1770 if not indices:
-> 1771     raise KeyError(f"Keyword {keyword!r} not found.")
   1773 try:
   1774     return indices[n]

KeyError: "Keyword 'BITPIX' not found."
```

Reply:
hmmm, I recently had issues around bitpix.... but this is i think a requirement of the input header....

Reply:
It's getting dropped somewhere here b/c the input header has it. not clear to me yet where it's getting dropped, though